### ISSUE TITLE
Colon's in HTML Class names

### ISSUE THREAD

**Issue by wesruv** [asciidoctor/asciidoctor#3767]
## TL;DR
Colon's are getting in class names, from what I can see, [I don't think it's valid HTML](https://www.w3.org/TR/CSS2/syndata.html) (see section 4.1.3).

But I can *definitely* say it's difficult to apply CSS styles to, and it's extremely difficult or impossible in JS to select something `:` in the selector.

As a front-end nerd, I'd suggest converting the `:` to `__` (double underscore) or `_` (single underscore). I think that'd make the part before and after the colon read separately, which I feel is the point of a `:` in the naming.

---------------------------

## Long version

On the project I'm working on we've decided to use [AsciiDoc Roles](https://asciidoctor-docs.netlify.app/asciidoc/1.5/attributes/role/) to denote certain important elements that need to be handled in a special way.

In this case we have an "Additional Resources" section, for content the reader might want to also read, the syntax we're using is something like this:

```adoc
[role="system:additional-resources"]
.Additional resources

* link:https://access.redhat.com/articles/3131341[Support Policies for RHEL High Availability Clusters - Microsoft Azure Virtual Machines as Cluster Members]
* link:https://access.redhat.com/articles/3071021[Support Policies for RHEL High Availability clusters - Transport Protocols]
* link:https://cloud.google.com/vpc/docs/vpc[VPC network overview]
```

It outputs this:
```html
<div class="ulist system:additional-resources">
  <h3 id="ulist_heading_0">Related content</h3>
  <ul aria-labelledby="ulist_heading_0">
    <li>
      <p><a href="https://access.redhat.com/articles/3131341">Support Policies for RHEL High Availability Clusters - Microsoft Azure Virtual Machines as Cluster Members</a></p>
    </li>
    <li>
      <p><a href="https://access.redhat.com/articles/3071021">Support Policies for RHEL High Availability clusters - Transport Protocols</a></p>
    </li>
    <li>
      <p><a href="https://cloud.google.com/vpc/docs/vpc">VPC network overview</a></p>
    </li>
  </ul>
</div>
```

I figured out I can style it with CSS if I escape the colon with a `\`, which I've never seen or heard of in my 17 years of writing CSS 
😂

e.g.
```css
.system\:additional-resources { /* Styles here */ }
```

I can't figure out a way to select it with JS, everything I've tried throws the error:
```
Uncaught DOMException: Failed to execute 'querySelector' on 'DocumentFragment': '.system:additional-resources' is not a valid selector.
```

In JS I've tried selecting with all of these and it throws the error:
* `.system:additional-resources`
* `.system\:additional-resources`
* `[class*=system:additional-resources]`
* `[class*=system\:additional-resources]`

**Comment by Mogztter** [comment]
Hey @wesruv 

As far as know there's little to no restriction on the characters that you can use in the `class` attribute:
https://www.w3.org/TR/2011/WD-html5-20110525/elements.html#classes


> As a front-end nerd, I'd suggest converting the : to __ (double underscore) or _ (single underscore). I think that'd make the part before and after the colon read separately, which I feel is the point of a : in the naming.

I don't think Asciidoctor should take the responsibility to replace symbols on the role attribute. It can potentially do more harm than good and arguably some users might find it confusing:

- why `:` is replaced by `_`?
- why `_` and not `-`?
- why `:` is replaced but not `*`?
- ...?

In my opinion, Asciidoctor should make sure that the HTML produced is valid, which is the case here.

I guess you should replace `:` by `_` or `__` in your AsciiDoc files or write a custom HTML template to replace `:`.

**Comment by wesruv** [comment]
Thanks, sorry for my confusion 😬 there was a good amount of it.

TL;DR I figured it out, but maybe consider removing chars that have to be escaped?

## My fix && confusion
Put some more heads together and found out that I need to double escape the colon in the querySelector call (which is bizarre to me). e.g.
`system\\:additional-resources`

So I'm guessing it is technically valid, but what a pain! The W3C spec talks about Unicode and ISO specs, which is where they lost me on "what's valid" in an HTML identifier.

Sorry, I originally thought the `system:` bit might have been an adoc convention, which is why I was suggesting `_`. Since it's just something my team did it doesn't make sense to make a special case. Changing to `_` or `__` is going to be a headache due to the amount of authors we have, so it wasn't were my mind went.


## Case for stripping chars that have to be escaped
They do something like this in Drupal (the CMS that I usually am working from), any back-end identifier that gets added to HTML as an ID or class name gets run through this function:
https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Component%21Utility%21Html.php/function/Html%3A%3AcleanCssIdentifier/8.2.x

If AsciiDoc roles are used for something on the backend, I think the same argument could apply, but I could see the argument that it should be left intact for authorship intent, despite the difficulty it causes in CSS/HTML.

Feel free to close this!

**Comment by mojavelinux** [comment]
The solution is to either not use colons in role names (if you find them hard to work with) or use the colon escaping feature of CSS. I don't see anything Asciidoctor should be doing differently here.
